airflow/models/taskinstance.py

```diff
@@ -1328,22 +1389,58 @@ def _update_ti_state_for_sensing(self, session=None):
 
     def _execute_task(self, context, task_copy):
         """Executes Task (optionally with a Timeout) and pushes Xcom results"""
+        # If the task has been deferred and is being executed due to a trigger,
+        # then we need to pick the right method to come back to, otherwise
+        # we go for the default execute
+        execute_callable = task_copy.execute
+        if self.next_method:
+            execute_callable = getattr(task_copy, self.next_method)
+            if self.next_kwargs:
+                execute_callable = partial(execute_callable, **self.next_kwargs)
         # If a timeout is specified for the task, make it fail
         # if it goes beyond
         if task_copy.execution_timeout:
             try:
                 with timeout(task_copy.execution_timeout.total_seconds()):
-                    result = task_copy.execute(context=context)
+                    result = execute_callable(context=context)
             except AirflowTaskTimeout:
                 task_copy.on_kill()
                 raise
         else:
-            result = task_copy.execute(context=context)
+            result = execute_callable(context=context)
         # If the task returns a result, push an XCom containing it
         if task_copy.do_xcom_push and result is not None:
             self.xcom_push(key=XCOM_RETURN_KEY, value=result)
         return result
 
+    @provide_session
+    def _defer_task(self, session, defer: TaskDeferred):
+        """
+        Marks the task as deferred and sets up the trigger that is needed
+        to resume it.
+        """
+        from airflow.models.trigger import Trigger
+
+        # First, make the trigger entry
+        trigger_row = Trigger.from_object(defer.trigger)
+        session.add(trigger_row)
+        session.commit()
```

Comment:
https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#database-session-handling

> If a function accepts a session parameter it should not commit the transaction itself. Session management is up to the caller.

(And provide session will commit for us automatically if it creates the session)

```suggestion
        session.flush()
```